### PR TITLE
[spirv] More support for matrix and struct type in init lists

### PR DESCRIPTION
**Pull request by antiagainst**
Now we can decompose matrices appeared in init lists. Also we can
decompose and construct structs from init lists.

### CODE REVIEW

File: tools/clang/lib/SPIRV/InitListHandler.cpp [hunk 1]
```diff
@@ -64,22 +73,81 @@ void InitListHandler::decompose(const Expr *expr) {
...
+        context, init, /*isarraw*/ false, /*OperatorLoc*/ {},
+        /*QualifierLoc*/ {}, /*TemplateKWLoc*/ {}, field,
+        DeclAccessPair::make(field, AS_none),
+        DeclarationNameInfo(field->getDeclName(), /*NameLoc*/ {}),
+        /*TemplateArgumentListInfo*/ nullptr, field->getType(),
+        init->getValueKind(), OK_Ordinary));
+  }
+
+  // Push in the reverse order
+  for (int32_t i = fields.size() - 1; i >= 0; --i)
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/InitListHandler.cpp]: you could replace the loop with:
```cpp
initializers.insert(initializers.end(), fields.rbegin(), fields.rend());
```

File: tools/clang/lib/SPIRV/InitListHandler.cpp [hunk 2]
```diff
@@ -28,12 +31,18 @@ uint32_t InitListHandler::process(const InitListExpr *expr) {
   scalars.clear();
 
   flatten(expr);
+  // Reverse the whole initializer list so we can manipulate the list at the
+  // tail of the vector. This is more efficient than using a deque.
+  std::reverse(std::begin(initializers), std::end(initializers));
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/InitListHandler.cpp]: Nice!

File: tools/clang/test/CodeGenSPIRV/type.struct.hlsl [hunk 3]
```diff
@@ -10,6 +12,9 @@
 // CHECK-NEXT: OpMemberName %T 1 "y"
 // CHECK-NEXT: OpMemberName %T 2 "z"
 
+// CHECK:      %N = OpTypeStruct
+struct N {};
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/type.struct.hlsl]: good one!

File: tools/clang/test/CodeGenSPIRV/var.init.struct.hlsl [hunk 4]
```diff
@@ -0,0 +1,91 @@
+// Run: %dxc -T vs_6_0 -E main
+
+struct S {
+    int3 a;
+    uint b;
+    float2x2 c;
+};
+
+struct T {
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/var.init.struct.hlsl]: +1 this is a great test case :)

File: tools/clang/lib/SPIRV/InitListHandler.cpp [hunk 5]
```diff
@@ -64,22 +73,81 @@ void InitListHandler::decompose(const Expr *expr) {
...
-      }
+      for (uint32_t i = 0; i < rowCount; ++i)
+        for (uint32_t j = 0; j < colCount; ++j) {
+          const uint32_t element =
+              theBuilder.createCompositeExtract(elemTypeId, mat, {i, j});
+          scalars.emplace_back(element, elemType);
+        }
     }
+  } else if (type->isStructureType()) {
+    assert("should already handled");
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/InitListHandler.cpp]: Good catch! That is true. I'll change this to `llvm_unreachable`.

File: tools/clang/lib/SPIRV/InitListHandler.cpp (hunk 1 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/InitListHandler.cpp]: +1

**Pull request merged**